### ISSUE TITLE
Issue while computing the `flow.forward` function

### ISSUE THREAD

**Issue by h2o64** [VincentStimper/normalizing-flows#27]
Hi,

I was wondering why there was a minus here https://github.com/VincentStimper/normalizing-flows/blob/0466e7f530b5091e60eb35f6ce00ce004fd8845f/normflows/core.py#L54 when computing the forward log jacobian ? Shouldn't it be a plus ? I made my own test case (see bellow) and the jacobian core test https://github.com/VincentStimper/normalizing-flows/blob/0466e7f530b5091e60eb35f6ce00ce004fd8845f/normflows/core_test.py#L58 fails for me.

```
# Libraries
import torch
import numpy as np
import normflows as nf
from tqdm import trange

# Train a flow
def train_flow(flow, base, target, batch_size=1024, lr=5e-4, n_iter=4096):
	# Optimizer
	optimizer = torch.optim.Adam(flow.parameters(), lr=lr)
	# Train the flow
	r = trange(n_iter, unit='step', desc='loss')
	for i in r:
	    # Reset the optimizer
	    optimizer.zero_grad()
	    # Sample the target
	    target_samples = target.sample((batch_size,))
	    # Estimate the loss
	    z, log_jac = flow.inverse_and_log_det(target_samples)
	    flow_loss = -(base.log_prob(z) + log_jac).mean()
	    # Backward pass
	    flow_loss.backward()
	    optimizer.step()
	    # Debug
	    r.set_description('loss = {}'.format(round(flow_loss.item(), 4)), refresh=True)

# Make the device
device = torch.device('cuda')

# Make the arguments
dim = 2

# Make the base
base = torch.distributions.MultivariateNormal(
    loc=torch.zeros((dim,), device=device),
    covariance_matrix=torch.eye(dim, device=device)
)
locs = torch.ones((2, dim), device=device)
locs[1] *= -1
target = torch.distributions.MixtureSameFamily(
	mixture_distribution=torch.distributions.Categorical(torch.ones((2,), device=device)),
	component_distribution=torch.distributions.MultivariateNormal(
	    loc=locs,
	    covariance_matrix=torch.stack([0.3 * torch.eye(dim, device=device)])
	)
)

# Make the flow
base_ = nf.distributions.DiagGaussian(dim, trainable=False)
flows = []
b = torch.Tensor([1 if i % 2 == 0 else 0 for i in range(2)])
for i in range(2):
    s = nf.nets.MLP([2, 2 * 2, 2], init_zeros=True)
    t = nf.nets.MLP([2, 2 * 2, 2], init_zeros=True)
    if i % 2 == 0:
        flows += [nf.flows.MaskedAffineFlow(b, t, s)]
    else:
        flows += [nf.flows.MaskedAffineFlow(1 - b, t, s)]
    flows += [nf.flows.ActNorm(2)]
flow = nf.NormalizingFlow(base, flows)
flow = flow.to(device)

# Launch training
train_flow(flow, base, target, batch_size=2048, lr=1e-3)

# Test bijectivity
z = base.sample(sample_shape=(2048,))
with torch.no_grad():
	x, log_jac_x = flow.forward_and_log_det(z)
	z_z, log_jac_z_z = flow.inverse_and_log_det(x)
with torch.no_grad():
	print('norm(z - z_z) = ', torch.linalg.norm((z - z_z).view((-1, 2)), dim=-1).mean())
	print('mse(log_jac_z_z + log_jac_x) = ', torch.mean(torch.square(log_jac_z_z + log_jac_x)))
	print('mse(log_jac_z_z - log_jac_x) = ', torch.mean(torch.square(log_jac_z_z - log_jac_x)))

# Display the push-forward
import matplotlib.pyplot as plt
res = 64
x_min, x_max, y_min, y_max = -3, 3, -3, 3
x = np.linspace(x_min, x_max, res)
y = np.linspace(x_min, x_max, res)
X, Y = np.meshgrid(x, y)
Z = np.stack([X,Y], axis=-1).reshape((-1,2))
Z = torch.from_numpy(Z).float().to(device)
Z_z, Z_z_log_jac = flow.inverse_and_log_det(Z)
log_prob_flow_forward = base.log_prob(Z_z) + Z_z_log_jac
log_prob_flow_forward = log_prob_flow_forward.reshape((res, res)).detach().cpu()
plt.contourf(X, Y, torch.exp(log_prob_flow_forward), 20, cmap='GnBu')
plt.xlabel('x')
plt.ylabel('y')
plt.title('Push-forward')
plt.show()
```

This code returns 
```
norm(z - z_z) =  tensor(2.3041e-07, device='cuda:0')
mse(log_jac_z_z + log_jac_x) =  tensor(1.3097, device='cuda:0')
mse(log_jac_z_z - log_jac_x) =  tensor(6.4206e-15, device='cuda:0')
```

**Comment by h2o64** [comment]
I think I understand why the assertion is not failing in `core_test.py `. The flow used is initialized with `init_zeros=True` making it the identity map
https://github.com/VincentStimper/normalizing-flows/blob/0466e7f530b5091e60eb35f6ce00ce004fd8845f/normflows/core_test.py#L27-L28
Because the identity map is its own inverse, the jacobians are the same.

**Comment by VincentStimper** [comment]
Hi @h2o64,

thanks for spotting this! I have added this method recently and haven't used it yet, which is why I did not notice it.

I made a [PR](https://github.com/VincentStimper/normalizing-flows/pull/28) solving this, feel free to have a look.

Best regards,
Vincent

**Comment by h2o64** [comment]
Looks good to me !

Thank you for your awesome package,
Louis
